### PR TITLE
Update the state with results from kevm-vm

### PR DESCRIPTION
**Pull request by anvacaru**
- Record `CallResult` message from `kevm-vm` back into the `StateManager`
- process `VMQuery`: `GetCode`

### CODE REVIEW

File: lib/runCall.js [hunk 1]
```diff
@@ -59,111 +61,287 @@ module.exports = function (opts, cb) {
...
+          case QUERY_NOT_SET: {
+            console.log('QUERYNOTSET')
+            break
+          }
         }
-        case GETBLOCKHASH: {
-          console.log('GETBLOCKHASH', data)
-          // TODO
-          break
+      } catch (err) {
```
Comment (ehildenb) [inline lib/runCall.js:84]: Let's localize this `catch` block to where the `try` is which is throwing the assertion error.

**Review by ehildenb — CHANGES_REQUESTED** [review]

**Comment by dwightguth** [comment]
As we said in the meeting, a code of "" in the CallResult means that the code was not modified by this transaction.

File: lib/runCall.js [hunk 2]
```diff
@@ -59,111 +61,287 @@ module.exports = function (opts, cb) {
...
+            console.log(err)
+          }
+          stateAccount.nonce = Buffer(nonce)
+          stateAccount.balance = Buffer(newBalance)
+          stateManager.putAccount(address, stateAccount, _callback)
+        })
+      },
+        function (_callback) {
+          stateManager.putContractCode(address, code, _callback)
+        },
```
Comment (ehildenb) [inline lib/runCall.js:246]: We should be checking if `code == ""`, and if so not update the contract code.

File: lib/runCall.js [hunk 3]
```diff
@@ -59,111 +61,287 @@ module.exports = function (opts, cb) {
...
     var address = getAccountObject.getAddress()
-    console.log('addressHex', Buffer(address).toString('hex'))
     stateManager.getAccount(new Buffer(address), function (err, fromAccount) {
-      account = createAccount(fromAccount)
+      if (fromAccount.isEmpty()) { // FIX THIS (createdAddress = returnCode)
+        if (salt) {
+          createdAddress = toAddress = ethUtil.generateAddress2(caller, salt, code)
+        } else {
+          createdAddress = toAddress = ethUtil.generateAddress(caller, new BN(caller.nonce).toArray())
+        }
```
Comment (ehildenb) [inline lib/runCall.js]: Get the `createdAddress` from the kevm-vm message not recompute it.

File: lib/runCall.js (hunk 2 above)
Comment (anvacaru) [reply lib/runCall.js:246]: Added a new condition to check the `code` length.

File: lib/runCall.js (hunk 1 above)
Comment (anvacaru) [reply lib/runCall.js:84]: Moved the `switch` statement outside of the `try` block.

File: lib/runCall.js [hunk 4]
```diff
@@ -59,111 +61,279 @@ module.exports = function (opts, cb) {
...
+    var returnCode = callResultObject.getReturncode()
+    var gasRemaining = new BN(callResultObject.getGasremaining())
+    var gasRefund = new BN(callResultObject.getGasrefund())
+    var error = callResultObject.getError()
+    var modifiedAccounts = callResultObject.getModifiedaccountsList()
+    var logEntries = callResultObject.getLogsList()
+    var logs = fromLogEntries(logEntries)
+    if (returnCode.length === 20) {
+      createdAddress = ethUtil.bufferToHex(Buffer.from(returnCode))
+    } else if (returnCode.length === 21) {
```
Comment (anvacaru) [inline lib/runCall.js:149]: The `createdAddress` is now taken from the `CallResult` message. There are cases in which the `returnCode` has an extra byte on the first position which results in an invalid address. I'm not sure why it happens and this should be considered as a temporary work around.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**